x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java

```diff
@@ -136,53 +148,276 @@ protected void readInternal(ByteBuffer b) throws IOException {
+        if (length == CodecUtil.footerLength() && isClone == false && position == fileInfo.length() - length) {
+            if (readChecksumFromFileInfo(b)) {
+                logger.trace("read footer of file [{}] at position [{}], bypassing all caches", fileInfo.physicalName(), position);
+                return;
+            }
+            assert b.remaining() == length;
+        }
+
+        logger.trace("readInternal: read [{}-{}] ([{}] bytes) from [{}]", position, position + length, length, this);
+
+        final Tuple<Long, Long> indexCacheMiss; // null if not a miss
+
+        // We prefer to use the index cache if the recovery is not done yet
+        if (directory.isRecoveryDone() == false) {
+            // We try to use the snapshot blob cache if:
+            // - the file is small enough to be fully cached in the blob cache
+            final boolean canBeFullyCached = fileInfo.length() <= BlobStoreCacheService.DEFAULT_CACHED_BLOB_SIZE * 2;
+            // - we're reading the first N bytes of the file
+            final boolean isStartOfFile = (position + length <= BlobStoreCacheService.DEFAULT_CACHED_BLOB_SIZE);
+
+            if (canBeFullyCached || isStartOfFile) {
+                final CachedBlob cachedBlob = directory.getCachedBlob(fileInfo.physicalName(), 0L, length);
+
+                if (cachedBlob == CachedBlob.CACHE_MISS || cachedBlob == CachedBlob.CACHE_NOT_READY) {
+                    // We would have liked to find a cached entry but we did not find anything: the cache on the disk will be requested so
+                    // we compute the regions of the file we would like to have the next time. The regions are expressed as tuples of
+                    // {start, end} ranges where positions are relative to the whole file.
+                    if (canBeFullyCached) {
+                        // if the index input is smaller than twice the size of the blob cache, it will be fully indexed
+                        indexCacheMiss = Tuple.tuple(0L, fileInfo.length());
+                    } else {
+                        // the index input is too large to fully cache, so just cache the initial range
+                        indexCacheMiss = Tuple.tuple(0L, (long) BlobStoreCacheService.DEFAULT_CACHED_BLOB_SIZE);
+                    }
+
+                    // We must fill in a cache miss even if CACHE_NOT_READY since the cache index is only created on the first put.
+                    // TODO TBD use a different trigger for creating the cache index and avoid a put in the CACHE_NOT_READY case.
+                } else {
+                    logger.trace(
+                        "reading [{}] bytes of file [{}] at position [{}] using index cache",
+                        length,
+                        fileInfo.physicalName(),
+                        position
+                    );
+                    stats.addIndexCacheBytesRead(cachedBlob.length());
```

Comment:
Sure; didn't need the timings for tests but it might be useful in future, see d53443e753b.

Reply:
I've reverted this change because it interferes with other assertions about how the stats are collected when we are concurrently fetching and reading data. It's probably fixable in a followup but since there's no immediate need for these timings I won't do it here.